diff -x '*.orig' -x '*.rej' -x .config --unified --recursive --text --new-file linux-6.12_unpatched/arch/x86/kvm/svm/svm.c linux-6.12/arch/x86/kvm/svm/svm.c
--- a/arch/x86/kvm/svm/svm.c
+++ b/arch/x86/kvm/svm/svm.c
@@ -3201,13 +3201,8 @@ static int svm_set_msr(struct kvm_vcpu *vcpu, struct msr_data *msr)
 		if (data & ~supported_de_cfg)
 			return 1;
 
-		/*
-		 * Don't let the guest change the host-programmed value.  The
-		 * MSR is very model specific, i.e. contains multiple bits that
-		 * are completely unknown to KVM, and the one bit known to KVM
-		 * is simply a reflection of hardware capabilities.
-		 */
-		if (!msr->host_initiated && data != svm->msr_decfg)
+		/* Don't allow the guest to change a bit, #GP */
+		if (!msr->host_initiated && (data ^ supported_de_cfg))
 			return 1;
 
 		svm->msr_decfg = data;
