diff --git a/PKGBUILD b/PKGBUILD
index f1e2c2c..76c8d75 100644
--- a/PKGBUILD
+++ b/PKGBUILD
@@ -326,7 +326,9 @@ build() {
     --enable-rutabaga-gfx
     --smbd=/usr/bin/smbd
     --with-coroutine=ucontext
+    --target-list=x86_64-linux-user,x86_64-softmmu
   )
+  # Note: The custom target-list will cause errors in packaging.
   local configure_static_options=(
     "${common_configure_options[@]}"
     --enable-attr
@@ -364,8 +366,9 @@ build() {
     --disable-xen
     --disable-zstd
     --static
+    --target-list=x86_64-linux-user
   )
-
+  # Note: The custom target-list will cause errors in packaging.
   (
     cd build-static
     ../$pkgbase-$pkgver/configure "${configure_static_options[@]}"
